cosmwasm/packages/std/src/imports.rs

```diff
@@ -187,6 +215,227 @@ impl Api for ExternalApi {
+            0 => Ok(true),
+            1 => Ok(false),
+            3 => Err(VerificationError::InvalidHashFormat),
+            4 => Err(VerificationError::InvalidSignatureFormat),
+            5 => Err(VerificationError::InvalidPubkeyFormat),
+            10 => Err(VerificationError::GenericErr),
+            error_code => Err(VerificationError::unknown_err(error_code)),
+        }
+    }
+
+    /// Recovers a public key from a message hash and a signature.
+    ///
+    /// This is required when working with Ethereum where public keys
+    /// are not stored on chain directly.
+    ///
+    /// `recovery_param` must be 0 or 1. The values 2 and 3 are unsupported by this implementation,
+    /// which is the same restriction as Ethereum has (https://github.com/ethereum/go-ethereum/blob/v1.9.25/internal/ethapi/api.go#L466-L469).
+    /// All other values are invalid.
+    ///
+    /// Returns the recovered pubkey in compressed form, which can be used
+    /// in secp256k1_verify directly.
+    fn secp256k1_recover_pubkey(
+        &self,
+        message_hash: &[u8],
+        signature: &[u8],
+        recover_param: u8,
+    ) -> Result<Vec<u8>, RecoverPubkeyError> {
+        let hash_send = build_region(message_hash);
+        let hash_send_ptr = &*hash_send as *const Region as u32;
+        let sig_send = build_region(signature);
+        let sig_send_ptr = &*sig_send as *const Region as u32;
+
+        let result =
+            unsafe { secp256k1_recover_pubkey(hash_send_ptr, sig_send_ptr, recover_param.into()) };
+        let error_code = from_high_half(result);
+        let pubkey_ptr = from_low_half(result);
+        match error_code {
+            0 => {
+                let pubkey = unsafe { consume_region(pubkey_ptr as *mut Region) };
+                Ok(pubkey)
+            }
+            3 => Err(RecoverPubkeyError::InvalidHashFormat),
+            4 => Err(RecoverPubkeyError::InvalidSignatureFormat),
+            6 => Err(RecoverPubkeyError::InvalidRecoveryParam),
+            error_code => Err(RecoverPubkeyError::unknown_err(error_code)),
+        }
+    }
+
+    /// EdDSA ed25519 implementation.
+    ///
+    /// This function verifies messages against a signature, with the public key of the signer,
+    /// using the ed25519 elliptic curve digital signature parametrization / algorithm.
+    ///
+    /// The maximum currently supported message length is 4096 bytes.
+    /// The signature and public key are in [Tendermint](https://docs.tendermint.com/v0.32/spec/blockchain/encoding.html#public-key-cryptography)
+    /// format:
+    /// - signature: raw ED25519 signature (64 bytes).
+    /// - public key: raw ED25519 public key (32 bytes).
+    fn ed25519_verify(
+        &self,
+        message: &[u8],
+        signature: &[u8],
+        public_key: &[u8],
+    ) -> Result<bool, VerificationError> {
+        let msg_send = build_region(message);
+        let msg_send_ptr = &*msg_send as *const Region as u32;
```

Comment:
Can you extract this into a function? Would save a lot of boilerplate in every function

Reply:
Yeah, just copied it like that from CosmWasm v1, didn't want to start refactoring their code

Reply:
I would say lets clean it up.. especially if it's the cosmwasm v1 code that we want to integrate, we'll spend a little time here and end up with better code going forwards